services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AutoProvisioningTest.java

```diff
@@ -219,4 +197,57 @@ public void testDeviceRegistrationIsRemovedWhenAutoProvisionFails(final VertxTes
                     ctx.completeNow();
                 }));
     }
+
+    private void testProvisionSucceeds(final VertxTestContext ctx, final boolean isGateway) throws CertificateEncodingException {
+        final TrustedCertificateAuthority trustedCA = TENANT_INFO.getTrustedCertificateAuthorities().get(0);
+        final JsonObject clientContext = new JsonObject().put(CredentialsConstants.FIELD_CLIENT_CERT,
+                CERT.getEncoded());
+
+        // GIVEN a tenant CA with auto-provisioning enabled
+        trustedCA.setAutoProvisioningEnabled(true);
+
+        if (isGateway) {
+            // The property auto-provision-as-gateway is set to true
+            trustedCA.setAutoProvisioningAsGatewayEnabled(true);
+        }
+
+        when(deviceManagementService.createDevice(eq(tenantId), any(), any(), any()))
+                .thenReturn(Future.succeededFuture(OperationResult.ok(HttpURLConnection.HTTP_OK, Id.of(deviceId),
+                        Optional.empty(), Optional.empty())));
+        when(credentialsManagementService.updateCredentials(eq(tenantId), eq(deviceId), any(), any(),
+                any())).thenReturn(Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_CREATED)));
+
+        // WHEN provisioning a device/gateway from a certificate
+        AutoProvisioning
+                .provisionIfEnabled(tenantId, SUBJECT_DN, clientContext, credentialsManagementService,
+                        deviceManagementService, tenantInformationService, NoopSpan.INSTANCE)
+                .onComplete(ctx.succeeding(result -> {
+                    ctx.verify(() -> {
+                        // VERIFY that that the device/gateway has been registered.
+                        final ArgumentCaptor<Device> deviceCaptor = ArgumentCaptor.forClass(Device.class);
+                        verify(deviceManagementService).createDevice(eq(tenantId), any(),
+                                deviceCaptor.capture(), any());
+
+                        if (isGateway) {
+                            // VERIFY that a gateway has been provisioned by checking the relevant property
+                            assertThat(deviceCaptor.getValue().getAuthorities()
+                                    .contains(RegistryManagementConstants.AUTHORITY_AUTO_PROVISIONING_ENABLED))
+                                    .isTrue();
+                        }
+
+                        //VERIFY that the correct credentials are stored
+                        verify(credentialsManagementService).updateCredentials(eq(tenantId), eq(deviceId), any(), any(),
```

Comment:
since this an external service being invoked, we should make sure that the credentials passed into this method are of the expected type and have the expected properties.